File: MemoryCore/dify-plugin/memory_tencentdb_dify/base_adapter.py
"""MemoryAdapterBase -- Abstract base class for all platform adapters.

This is the Python counterpart to the TypeScript ``MemoryAdapterBase`` in
``src/adapters/sdk/base-adapter.ts``.

New platforms extend this class and implement only 4 abstract methods:

1. ``format_recall_result(result)`` -- Format memory recall output for the
   platform's prompt injection convention.
2. ``get_tool_definitions()`` -- Return tool schemas the platform registers
   with its LLM.
3. ``format_tool_result(tool_name, raw_result)`` -- Format tool call output
   for the platform's expected response format.
4. ``normalize_messages(raw_messages, context)`` -- Convert platform-specific
   message format into the standard ``ConversationMessage`` list.

The base class handles:
- Gateway connection and health checking
- Circuit breaker (pauses calls after N consecutive failures)
- Recall (parallel L1 + L3 + L2 fetch via MemoryGatewayClient)
- Capture (conversation add via MemoryGatewayClient)
- Search (memory + conversation search via MemoryGatewayClient)
- Session management

Usage::

    class DifyAdapter(MemoryAdapterBase):
        platform_name = "dify"
        # implement 4 abstract methods...

    adapter = DifyAdapter()
    adapter.initialize({"gateway": {"endpoint": "http://127.0.0.1:8420"}})
    prepend, append = adapter.recall("user query", "session-1")
    adapter.capture(messages, "session-1")
"""

from __future__ import annotations

import abc
import logging
import time
from typing import Any, Dict, List, Optional, Tuple, Union

from .gateway_client import MemoryGatewayClient
from .types import (
    AdapterConfig,
    CaptureResult,
    ConversationMessage,
    FormattedRecallResult,
    RecallResult,
    SearchResult,
    TenancyConfig,
    ToolDefinition,
)

logger = logging.getLogger(__name__)

__all__ = [
    "MemoryAdapterBase",
    "DEFAULT_MEMORY_SEARCH_TOOL",
    "DEFAULT_CONVERSATION_SEARCH_TOOL",
    "DEFAULT_READ_SCENE_TOOL",
    "BREAKER_THRESHOLD",
    "BREAKER_COOLDOWN_MS",
]

# ============================
# Circuit breaker constants
# ============================

BREAKER_THRESHOLD = 5
BREAKER_COOLDOWN_MS = 60_000


# ============================
# Default tool definitions
# ============================

DEFAULT_MEMORY_SEARCH_TOOL = ToolDefinition(
    name="tdai_memory_search",
    label="Memory Search",
    description=(
        "Search structured memories (L1). Returns relevant memory fragments "
        "about user preferences, past events, rules, and facts."
    ),
    parameters={
        "type": "object",
        "properties": {
            "query": {
                "type": "string",
                "description": "Search query text (natural language).",
            },
            "limit": {
                "type": "number",
                "description": "Max results to return (default: 5).",
            },
            "type": {
                "type": "string",
                "description": "Filter by memory type.",
            },
        },
        "required": ["query"],
    },
)

DEFAULT_CONVERSATION_SEARCH_TOOL = ToolDefinition(
    name="tdai_conversation_search",
    label="Conversation Search",
    description=(
        "Search raw conversation history (L0). Returns original messages "
        "with timestamps."
    ),
    parameters={
        "type": "object",
        "properties": {
            "query": {
                "type": "string",
                "description": "Search query text.",
            },
            "limit": {
                "type": "number",
                "description": "Max results (default: 5).",
            },
            "session_key": {
                "type": "string",
                "description": "Filter by session ID.",
            },
        },
        "required": ["query"],
    },
)

DEFAULT_READ_SCENE_TOOL = ToolDefinition(
    name="tdai_read_scene",
    label="Read Scene",
    description=(
        "Read a scene block's full content by its name. "
        "Use when you see a scene listed in Scene Navigation."
    ),
    parameters={
        "type": "object",
        "properties": {
            "scene_id": {
                "type": "string",
                "description": "Scene file name (e.g. 'travel-plan.md').",
            },
        },
        "required": ["scene_id"],
    },
)


class MemoryAdapterBase(abc.ABC):
    """Abstract base class for all platform adapters.

    Subclasses implement 4 abstract methods and optionally override the
    lifecycle hooks (``on_gateway_ready``, ``on_gateway_unavailable``,
    ``on_recall_error``, ``on_capture_error``). All Gateway communication,
    health checking, circuit breaking, and error handling are handled here.
    """

    #: Platform identifier (e.g. ``"dify"``, ``"hermes"``). Set by subclass.
    platform_name: str = ""

    def __init__(self) -> None:
        self._client: Optional[MemoryGatewayClient] = None
        self._config: Optional[AdapterConfig] = None
        self._tenancy: Dict[str, str] = {
            "team_id": "default",
            "agent_id": "default",
            "user_id": "default",
        }
        self._session_id: str = ""

        # Circuit breaker state
        self._consecutive_failures: int = 0
        self._breaker_open_until: float = 0.0

    # -- Properties ----------------------------------------------------------

    @property
    def is_initialized(self) -> bool:
        """Whether :meth:`initialize` has been called successfully."""
        return self._client is not None

    @property
    def session_id(self) -> str:
        """The current session ID for capture operations."""
        return self._session_id

    # ============================
    # Lifecycle
    # ============================

    def initialize(self, config: AdapterConfig) -> None:
        """Initialize the adapter with configuration.

        Creates the Gateway client and performs a best-effort health check.
        Does not raise on health check failure -- the Gateway might come up
        later.

        Args:
            config: Full adapter configuration. See :class:`AdapterConfig`.
        """
        self._config = config
        gateway = config.get("gateway", {})
        tenancy = config.get("tenancy") or {}

        self._tenancy = {
            "team_id": tenancy.get("team_id", "default") or "default",
            "agent_id": tenancy.get("agent_id", "default") or "default",
            "user_id": tenancy.get("user_id", "default") or "default",
        }

        self._client = MemoryGatewayClient(
            endpoint=gateway.get("endpoint", "http://127.0.0.1:8420"),
            api_key=gateway.get("api_key"),
            service_id=gateway.get("service_id", "default"),
            timeout_ms=gateway.get("timeout_ms", 10_000),
            reject_unauthorized=gateway.get("reject_unauthorized", True),
        )

        # Best-effort health check (non-blocking on failure)
        health = self._client.health(timeout_ms=3_000)
        status = health.get("status", "unknown")
        if status in ("ok", "degraded"):
            self.on_gateway_ready()
        else:
            self.on_gateway_unavailable(status)

    def set_session_id(self, session_id: str) -> None:
        """Set the current session ID for capture operations.

        Args:
            session_id: The session identifier.
        """
        self._session_id = session_id

    def shutdown(self) -> None:
        """Shutdown the adapter.

        The base class has no persistent resources to close. Subclasses can
        override for platform-specific cleanup.
        """
        pass

    # ============================
    # Core capabilities (implemented by base class)
    # ============================

    def recall(
        self,
        query: str,
        session_id: Optional[str] = None,
    ) -> Tuple[str, str]:
        """Recall memories for the current user query.

        Performs parallel L1 (structured memories) + L3 (persona) + L2 (scene
        navigation) fetch from the Gateway, then delegates formatting to the
        platform-specific :meth:`format_recall_result` method.

        Returns empty strings on failure -- never raises.

        Args:
            query: The user's query text.
            session_id: Optional session override. Falls back to the
                session set via :meth:`set_session_id`.

        Returns:
            A tuple of ``(prepend_context, append_system_context)`` strings.
        """
        if not query or self._client is None or self._is_breaker_open():
            return "", ""

        effective_session = session_id or self._session_id
        start_ms = self._now_ms()

        try:
            memories, persona, scenes = self._client.recall(
                query,
                self._tenancy,
                options={
                    "max_results": (self._config or {}).get("recall_max_results", 5),
                    "include_persona": (self._config or {}).get("recall_include_persona", True),
                    "include_scene_nav": (self._config or {}).get("recall_include_scene_nav", True),
                },
            )

            latency_ms = self._now_ms() - start_ms
            result = RecallResult(
                prepend_context="",
                append_system_context="",
                memories=memories,
                persona=persona,
                scenes=scenes,
                latency_ms=latency_ms,
            )

            self._record_success()

            formatted = self.format_recall_result(result)
            return (
                formatted.get("prepend_context", "") or "",
                formatted.get("append_system_context", "") or "",
            )
        except Exception as err:
            self._record_failure()
            self.on_recall_error(err)
            return "", ""

    def capture(
        self,
        raw_messages: Any,
        session_id: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> CaptureResult:
        """Capture conversation messages.

        Normalizes platform-specific messages via :meth:`normalize_messages`,
        then sends them to the Gateway for L0 recording.

        Returns a failure result on error -- never raises.

        Args:
            raw_messages: Platform-specific message format.
            session_id: Optional session override.
            context: Optional context dict passed to :meth:`normalize_messages`.

        Returns:
            A :class:`CaptureResult` with capture status.
        """
        if self._client is None or self._is_breaker_open():
            return CaptureResult(captured_count=0, success=False, error="circuit breaker open")

        if not (self._config or {}).get("capture_enabled", True):
            return CaptureResult(captured_count=0, success=True)

        try:
            messages = self.normalize_messages(raw_messages, context)
            if not messages:
                return CaptureResult(captured_count=0, success=True)

            effective_session = session_id or self._session_id
            captured_count, success = self._client.capture(
                messages,
                effective_session,
                self._tenancy,
            )

            if success:
                self._record_success()
            else:
                self._record_failure()

            return CaptureResult(captured_count=captured_count, success=success)
        except Exception as err:
            self._record_failure()
            error_msg = str(err)
            self.on_capture_error(error_msg)
            return CaptureResult(captured_count=0, success=False, error=error_msg)

    def search_memories(
        self,
        query: str,
        limit: Optional[int] = None,
        type_filter: Optional[str] = None,
    ) -> SearchResult:
        """Search L1 structured memories.

        Args:
            query: Search query text.
            limit: Max results to return (default: 5).
            type_filter: Filter by memory type.

        Returns:
            A :class:`SearchResult`. Returns empty result on failure.
        """
        if not query or self._client is None or self._is_breaker_open():
            return SearchResult(text="No results.", total=0)

        try:
            options: Dict[str, Any] = {}
            if limit is not None:
                options["limit"] = limit
            if type_filter is not None:
                options["type"] = type_filter

            items, total = self._client.search_memories(
                query,
                self._tenancy,
                options=options if options else None,
            )
            self._record_success()

            if not items:
                text = "No memories found for this query."
            else:
                text = "\n".join(f"- [{m.type}] {m.content}" for m in items)

            return SearchResult(text=text, total=total, items=items)
        except Exception as err:
            self._record_failure()
            return SearchResult(text=f"Search failed: {err}", total=0)

    def search_conversations(
        self,
        query: str,
        limit: Optional[int] = None,
        session_id: Optional[str] = None,
    ) -> SearchResult:
        """Search L0 raw conversations.

        Args:
            query: Search query text.
            limit: Max results to return (default: 5).
            session_id: Filter by session ID.

        Returns:
            A :class:`SearchResult`. Returns empty result on failure.
        """
        if not query or self._client is None or self._is_breaker_open():
            return SearchResult(text="No results.", total=0)

        try:
            options: Dict[str, Any] = {}
            if limit is not None:
                options["limit"] = limit
            if session_id is not None:
                options["session_id"] = session_id

            items, total = self._client.search_conversations(
                query,
                self._tenancy,
                options=options if options else None,
            )
            self._record_success()

            if not items:
                text = "No conversations found for this query."
            else:
                text = "\n".join(f"[{m.type}] {m.content}" for m in items)

            return SearchResult(text=text, total=total, items=items)
        except Exception as err:
            self._record_failure()
            return SearchResult(text=f"Search failed: {err}", total=0)

    def read_scene(self, scene_id: str) -> str:
        """Read a scene block by path.

        Args:
            scene_id: Scene file name (with or without ``.md`` extension).

        Returns:
            The scene content text, or an error message on failure.
        """
        if not scene_id or self._client is None or self._is_breaker_open():
            return "Scene not available."

        try:
            path = scene_id if scene_id.endswith(".md") else f"{scene_id}.md"
            content = self._client.read_scene(path, self._tenancy)
            self._record_success()
            return content or f"Scene '{scene_id}' is empty or not found."
        except Exception as err:
            self._record_failure()
            return f"Failed to read scene: {err}"

    def handle_tool_call(
        self,
        tool_name: str,
        args: Dict[str, Any],
    ) -> str:
        """Handle a tool call by name. Dispatches to the appropriate method.

        Supports both the generic (``tdai_*``) and provider-prefixed
        (``memory_tencentdb_*``) tool name conventions.

        Args:
            tool_name: The tool name to dispatch.
            args: Tool call arguments.

        Returns:
            The formatted tool result as a string.
        """
        try:
            raw_result: Union[SearchResult, str]

            if tool_name in ("tdai_memory_search", "memory_tencentdb_memory_search"):
                raw_result = self.search_memories(
                    str(args.get("query", "")),
                    limit=args.get("limit") if isinstance(args.get("limit"), (int, float)) else None,
                    type_filter=args.get("type") if isinstance(args.get("type"), str) else None,
                )
            elif tool_name in ("tdai_conversation_search", "memory_tencentdb_conversation_search"):
                raw_result = self.search_conversations(
                    str(args.get("query", "")),
                    limit=args.get("limit") if isinstance(args.get("limit"), (int, float)) else None,
                    session_id=args.get("session_key") if isinstance(args.get("session_key"), str) else None,
                )
            elif tool_name in ("tdai_read_scene", "memory_tencentdb_read_scene"):
                raw_result = self.read_scene(str(args.get("scene_id", "")))
            else:
                import json
                return json.dumps({"error": f"Unknown tool: {tool_name}"})

            return self.format_tool_result(tool_name, raw_result)
        except Exception as err:
            import json
            return json.dumps({"error": f"Tool call failed: {err}"})

    def get_tool_definitions(self) -> List[ToolDefinition]:
        """Get the default tool definitions.

        Returns the three default tools (memory search, conversation search,
        scene read). Platforms can override to customize.

        Returns:
            A list of :class:`ToolDefinition` instances.
        """
        return [
            DEFAULT_MEMORY_SEARCH_TOOL,
            DEFAULT_CONVERSATION_SEARCH_TOOL,
            DEFAULT_READ_SCENE_TOOL,
        ]

    # ============================
    # Circuit breaker
    # ============================

    def _is_breaker_open(self) -> bool:
        """Check whether the circuit breaker is currently open."""
        if self._consecutive_failures < BREAKER_THRESHOLD:
            return False
        if self._now_ms() >= self._breaker_open_until:
            self._consecutive_failures = 0
            return False
        return True

    def _record_success(self) -> None:
        """Reset the circuit breaker failure counter."""
        self._consecutive_failures = 0

    def _record_failure(self) -> None:
        """Increment the failure counter and trip the breaker if threshold reached."""
        self._consecutive_failures += 1
        if self._consecutive_failures >= BREAKER_THRESHOLD:
            self._breaker_open_until = self._now_ms() + BREAKER_COOLDOWN_MS
            logger.warning(
                "%s circuit breaker tripped after %d consecutive failures. "
                "Pausing Gateway calls for %dms.",
                self.platform_name or "adapter",
                self._consecutive_failures,
                BREAKER_COOLDOWN_MS,
            )

    @staticmethod
    def _now_ms() -> float:
        """Current monotonic time in milliseconds."""
        return time.monotonic() * 1000.0

    # ============================
    # Overridable hooks (platform-specific)
    # ============================

    def on_gateway_ready(self) -> None:
        """Called when the Gateway becomes reachable after :meth:`initialize`.

        Override in subclass for platform-specific logging.
        """
        pass

    def on_gateway_unavailable(self, status: str) -> None:
        """Called when the Gateway is not reachable at :meth:`initialize`.

        Override in subclass for platform-specific fallback.

        Args:
            status: The health check status string.
        """
        pass

    def on_recall_error(self, err: Exception) -> None:
        """Called when a recall operation fails.

        Override in subclass for platform-specific error handling.

        Args:
            err: The exception that caused the failure.
        """
        pass

    def on_capture_error(self, error_msg: str) -> None:
        """Called when a capture operation fails.

        Override in subclass for platform-specific error handling.

        Args:
            error_msg: The error message string.
        """
        pass

    # ============================
    # Abstract methods (must be implemented by each platform)
    # ============================

    @abc.abstractmethod
    def format_recall_result(self, result: RecallResult) -> FormattedRecallResult:
        """Format a recall result for this platform's prompt injection.

        Different platforms have different conventions for how memory context
        is injected into prompts. This method lets each platform control the
        exact format.

        Args:
            result: The raw :class:`RecallResult` from the Gateway.

        Returns:
            A dict with optional ``prepend_context`` and
            ``append_system_context`` keys.
        """
        ...

    @abc.abstractmethod
    def format_tool_result(
        self,
        tool_name: str,
        raw_result: Union[SearchResult, str],
    ) -> str:
        """Format tool call output for the platform's expected response format.

        Args:
            tool_name: The name of the tool that was called.
            raw_result: The raw :class:`SearchResult` or string result.

        Returns:
            The formatted result as a string.
        """
        ...

    @abc.abstractmethod
    def normalize_messages(
        self,
        raw_messages: Any,
        context: Optional[Dict[str, Any]] = None,
    ) -> List[ConversationMessage]:
        """Convert platform-specific message format into standard messages.

        Different platforms represent conversations differently. This method
        normalizes them into the standard :class:`ConversationMessage` list.

        Args:
            raw_messages: Platform-specific message format.
            context: Optional context dict.

        Returns:
            A list of :class:`ConversationMessage` instances.
        """
        ...

File: MemoryCore/src/adapters/sdk/base-adapter.ts
/**
 * MemoryAdapterBase — Abstract base class for all platform adapters.
 *
 * New platforms extend this class and implement only 4 abstract methods:
 *
 *   1. `formatRecallResult(result)` — Format memory recall output for the
 *      platform's prompt injection convention.
 *   2. `getToolDefinitions()` — Return tool schemas the platform registers
 *      with its LLM.
 *   3. `formatToolResult(toolName, rawResult)` — Format tool call output
 *      for the platform's expected response format.
 *   4. `normalizeMessages(rawMessages)` — Convert platform-specific message
 *      format into the standard `ConversationMessage[]`.
 *
 * The base class handles:
 *   - Gateway connection and health checking
 *   - Circuit breaker (pauses calls after N consecutive failures)
 *   - Recall (parallel L1 + L3 + L2 fetch via GatewayClient)
 *   - Capture (conversation add via GatewayClient)
 *   - Search (memory + conversation search via GatewayClient)
 *   - Session management
 *
 * Usage:
 *   ```typescript
 *   class ClaudeCodeAdapter extends MemoryAdapterBase {
 *     readonly platformName = "claude-code";
 *     // implement 4 abstract methods...
 *   }
 *
 *   const adapter = new ClaudeCodeAdapter();
 *   await adapter.initialize({ gateway: { endpoint: "http://127.0.0.1:8420" } });
 *   const recall = await adapter.recall("user query", "session-1");
 *   await adapter.capture(messages, "session-1");
 *   ```
 */

import { MemoryGatewayClient } from "./gateway-client.js";
import type {
  AdapterConfig,
  ConversationMessage,
  IPlatformAdapter,
  RecallResult,
  CaptureResult,
  SearchResult,
  ToolDefinition,
  TenancyConfig,
} from "./types.js";

// ============================
// Circuit breaker constants
// ============================

export const BREAKER_THRESHOLD = 5;
export const BREAKER_COOLDOWN_MS = 60_000;

// ============================
// Default tool definitions
// ============================

/** Default memory search tool definition. Platforms can customize. */
export const DEFAULT_MEMORY_SEARCH_TOOL: ToolDefinition = {
  name: "tdai_memory_search",
  label: "Memory Search",
  description:
    "Search structured memories (L1). Returns relevant memory fragments about " +
    "user preferences, past events, rules, and facts.",
  parameters: {
    type: "object",
    properties: {
      query: { type: "string", description: "Search query text (natural language)." },
      limit: { type: "number", description: "Max results to return (default: 5)." },
      type: { type: "string", description: "Filter by memory type." },
    },
    required: ["query"],
  },
};

/** Default conversation search tool definition. */
export const DEFAULT_CONVERSATION_SEARCH_TOOL: ToolDefinition = {
  name: "tdai_conversation_search",
  label: "Conversation Search",
  description:
    "Search raw conversation history (L0). Returns original messages with timestamps.",
  parameters: {
    type: "object",
    properties: {
      query: { type: "string", description: "Search query text." },
      limit: { type: "number", description: "Max results (default: 5)." },
      session_key: { type: "string", description: "Filter by session ID." },
    },
    required: ["query"],
  },
};

/** Default scene read tool definition. */
export const DEFAULT_READ_SCENE_TOOL: ToolDefinition = {
  name: "tdai_read_scene",
  label: "Read Scene",
  description:
    "Read a scene block's full content by its name. " +
    "Use when you see a scene listed in Scene Navigation.",
  parameters: {
    type: "object",
    properties: {
      scene_id: { type: "string", description: "Scene file name (e.g. 'travel-plan.md')." },
    },
    required: ["scene_id"],
  },
};

// ============================
// MemoryAdapterBase
// ============================

export abstract class MemoryAdapterBase implements IPlatformAdapter {
  abstract readonly platformName: string;

  protected client!: MemoryGatewayClient;
  protected config!: AdapterConfig;
  protected tenancy!: Required<TenancyConfig>;
  protected sessionId: string = "";

  // Circuit breaker state
  private consecutiveFailures = 0;
  private breakerOpenUntil = 0;

  // ============================
  // Properties
  // ============================

  /** Whether `initialize()` has been called successfully. */
  get isInitialized(): boolean {
    return this.client !== undefined;
  }

  /** The current session ID for capture operations. */
  get currentSessionId(): string {
    return this.sessionId;
  }

  // ============================
  // Lifecycle
  // ============================

  /**
   * Initialize the adapter with configuration.
   * Creates the Gateway client and performs a health check.
   */
  async initialize(config: AdapterConfig): Promise<void> {
    this.config = config;
    this.tenancy = {
      teamId: config.tenancy?.teamId ?? "default",
      agentId: config.tenancy?.agentId ?? "default",
      userId: config.tenancy?.userId ?? "default",
    };

    this.client = new MemoryGatewayClient({
      endpoint: config.gateway.endpoint,
      apiKey: config.gateway.apiKey,
      serviceId: config.gateway.serviceId,
      timeoutMs: config.gateway.timeoutMs,
      rejectUnauthorized: config.gateway.rejectUnauthorized,
    });

    // Best-effort health check (non-blocking on failure)
    const health = await this.client.health();
    if (health.status === "ok" || health.status === "degraded") {
      this.onGatewayReady();
    } else {
      // Don't throw — the Gateway might come up later
      this.onGatewayUnavailable(health.status);
    }
  }

  /**
   * Set the current session ID for capture operations.
   */
  setSessionId(sessionId: string): void {
    this.sessionId = sessionId;
  }

  /**
   * Shutdown the adapter. Subclasses can override for cleanup.
   */
  async shutdown(): Promise<void> {
    // Base class has no persistent resources to close.
    // Subclasses can override for platform-specific cleanup.
  }

  // ============================
  // Core capabilities (implemented by base class)
  // ============================

  /**
   * Recall memories for the current user query.
   *
   * Performs parallel L1 (structured memories) + L3 (persona) + L2 (scene
   * navigation) fetch from the Gateway, then delegates formatting to the
   * platform-specific `formatRecallResult()` method.
   *
   * Returns empty strings on failure — never throws.
   */
  async recall(
    query: string,
    sessionId?: string,
  ): Promise<{ prependContext: string; appendSystemContext: string }> {
    if (!query || !this.client || this.isBreakerOpen()) {
      return { prependContext: "", appendSystemContext: "" };
    }

    const effectiveSession = sessionId ?? this.sessionId;
    const startMs = Date.now();

    try {
      const { memories, persona, scenes } = await this.client.recall(
        query,
        this.tenancy,
        {
          maxResults: this.config.recallMaxResults ?? 5,
          includePersona: this.config.recallIncludePersona ?? true,
          includeSceneNav: this.config.recallIncludeSceneNav ?? true,
        },
      );

      const latencyMs = Date.now() - startMs;
      const result: RecallResult = {
        prependContext: "",
        appendSystemContext: "",
        memories,
        persona,
        scenes,
        latencyMs,
      };

      this.recordSuccess();

      // Delegate formatting to the platform implementation
      const formatted = this.formatRecallResult(result);
      return {
        prependContext: formatted.prependContext ?? "",
        appendSystemContext: formatted.appendSystemContext ?? "",
      };
    } catch (err) {
      this.recordFailure();
      this.onRecallError(err instanceof Error ? err : new Error(String(err)));
      return { prependContext: "", appendSystemContext: "" };
    }
  }

  /**
   * Capture conversation messages.
   *
   * Normalizes platform-specific messages via `normalizeMessages()`, then
   * sends them to the Gateway for L0 recording.
   *
   * Returns failure result on error — never throws.
   */
  async capture(
    rawMessages: unknown,
    sessionId?: string,
    context?: Record<string, unknown>,
  ): Promise<CaptureResult> {
    if (!this.client || this.isBreakerOpen()) {
      return { capturedCount: 0, success: false, error: "circuit breaker open" };
    }

    if (!(this.config.captureEnabled ?? true)) {
      return { capturedCount: 0, success: true };
    }

    try {
      const messages = this.normalizeMessages(rawMessages, context);
      if (messages.length === 0) {
        return { capturedCount: 0, success: true };
      }

      const effectiveSession = sessionId ?? this.sessionId;
      const result = await this.client.capture(
        messages,
        effectiveSession,
        this.tenancy,
      );

      if (result.success) {
        this.recordSuccess();
      } else {
        this.recordFailure();
      }

      return {
        capturedCount: result.capturedCount,
        success: result.success,
      };
    } catch (err) {
      this.recordFailure();
      const errorMsg = err instanceof Error ? err.message : String(err);
      this.onCaptureError(errorMsg);
      return { capturedCount: 0, success: false, error: errorMsg };
    }
  }

  /**
   * Search L1 structured memories.
   */
  async searchMemories(
    query: string,
    options?: { limit?: number; type?: string },
  ): Promise<SearchResult> {
    if (!query || !this.client || this.isBreakerOpen()) {
      return { text: "No results.", total: 0, items: [] };
    }

    try {
      const { items, total } = await this.client.searchMemories(
        query,
        this.tenancy,
        options,
      );
      this.recordSuccess();

      const text = items.length === 0
        ? "No memories found for this query."
        : items.map((m) => `- [${m.type}] ${m.content}`).join("\n");

      const result: SearchResult = { text, total, items };
      return result;
    } catch (err) {
      this.recordFailure();
      return { text: `Search failed: ${err instanceof Error ? err.message : String(err)}`, total: 0, items: [] };
    }
  }

  /**
   * Search L0 raw conversations.
   */
  async searchConversations(
    query: string,
    options?: { limit?: number; sessionId?: string },
  ): Promise<SearchResult> {
    if (!query || !this.client || this.isBreakerOpen()) {
      return { text: "No results.", total: 0, items: [] };
    }

    try {
      const { items, total } = await this.client.searchConversations(
        query,
        this.tenancy,
        options,
      );
      this.recordSuccess();

      const text = items.length === 0
        ? "No conversations found for this query."
        : items.map((m) => `[${m.type}] ${m.content}`).join("\n");

      const result: SearchResult = { text, total, items };
      return result;
    } catch (err) {
      this.recordFailure();
      return { text: `Search failed: ${err instanceof Error ? err.message : String(err)}`, total: 0, items: [] };
    }
  }

  /**
   * Read a scene block by path.
   */
  async readScene(sceneId: string): Promise<string> {
    if (!sceneId || !this.client || this.isBreakerOpen()) {
      return "Scene not available.";
    }

    try {
      const path = sceneId.endsWith(".md") ? sceneId : `${sceneId}.md`;
      const content = await this.client.readScene(path, this.tenancy);
      this.recordSuccess();
      return content || `Scene '${sceneId}' is empty or not found.`;
    } catch (err) {
      this.recordFailure();
      return `Failed to read scene: ${err instanceof Error ? err.message : String(err)}`;
    }
  }

  /**
   * Handle a tool call by name. Dispatches to the appropriate method.
   */
  async handleToolCall(
    toolName: string,
    args: Record<string, unknown>,
  ): Promise<string> {
    try {
      let rawResult: SearchResult | string;

      switch (toolName) {
        case "tdai_memory_search":
        case "memory_tencentdb_memory_search": {
          rawResult = await this.searchMemories(
            String(args.query ?? ""),
            {
              limit: typeof args.limit === "number" ? args.limit : undefined,
              type: typeof args.type === "string" ? args.type : undefined,
            },
          );
          break;
        }
        case "tdai_conversation_search":
        case "memory_tencentdb_conversation_search": {
          rawResult = await this.searchConversations(
            String(args.query ?? ""),
            {
              limit: typeof args.limit === "number" ? args.limit : undefined,
              sessionId: typeof args.session_key === "string" ? args.session_key : undefined,
            },
          );
          break;
        }
        case "tdai_read_scene":
        case "memory_tencentdb_read_scene": {
          rawResult = await this.readScene(String(args.scene_id ?? ""));
          break;
        }
        default:
          return JSON.stringify({ error: `Unknown tool: ${toolName}` });
      }

      return this.formatToolResult(toolName, rawResult);
    } catch (err) {
      return JSON.stringify({
        error: `Tool call failed: ${err instanceof Error ? err.message : String(err)}`,
      });
    }
  }

  /**
   * Get the default tool definitions (memory search, conversation search,
   * scene read). Platforms can override to customize.
   */
  getToolDefinitions(): ToolDefinition[] {
    return [
      DEFAULT_MEMORY_SEARCH_TOOL,
      DEFAULT_CONVERSATION_SEARCH_TOOL,
      DEFAULT_READ_SCENE_TOOL,
    ];
  }

  // ============================
  // Circuit breaker
  // ============================

  protected isBreakerOpen(): boolean {
    if (this.consecutiveFailures < BREAKER_THRESHOLD) return false;
    if (Date.now() >= this.breakerOpenUntil) {
      this.consecutiveFailures = 0;
      return false;
    }
    return true;
  }

  protected recordSuccess(): void {
    this.consecutiveFailures = 0;
  }

  protected recordFailure(): void {
    this.consecutiveFailures++;
    if (this.consecutiveFailures >= BREAKER_THRESHOLD) {
      this.breakerOpenUntil = Date.now() + BREAKER_COOLDOWN_MS;
    }
  }

  // ============================
  // Overridable hooks (platform-specific)
  // ============================

  /** Called when the Gateway becomes reachable after initialize(). */
  protected onGatewayReady(): void {
    // Override in subclass for platform-specific logging
  }

  /** Called when the Gateway is not reachable at initialize(). */
  protected onGatewayUnavailable(_status: string): void {
    // Override in subclass for platform-specific fallback
  }

  /** Called when a recall operation fails. */
  protected onRecallError(_err: Error): void {
    // Override in subclass for platform-specific error handling
  }

  /** Called when a capture operation fails. */
  protected onCaptureError(_errorMsg: string): void {
    // Override in subclass for platform-specific error handling
  }

  // ============================
  // Abstract methods (must be implemented by each platform)
  // ============================

  abstract formatRecallResult(result: RecallResult): {
    prependContext?: string;
    appendSystemContext?: string;
  };

  abstract formatToolResult(
    toolName: string,
    rawResult: SearchResult | string,
  ): string;

  abstract normalizeMessages(
    rawMessages: unknown,
    context?: Record<string, unknown>,
  ): ConversationMessage[];
}
